frame/evm/precompile/clear-storage/src/lib.rs

```diff
@@ -0,0 +1,110 @@
+// SPDX-License-Identifier: Apache-2.0
+// This file is part of Frontier.
+//
+// Copyright (c) 2020-2022 Parity Technologies (UK) Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+//#![deny(unused_crate_dependencies)]
+
+extern crate alloc;
+
+pub const ARRAY_LIMIT: u32 = 1_000;
+
+use core::marker::PhantomData;
+use fp_evm::ACCOUNT_STORAGE_PROOF_SIZE;
+use pallet_evm::AddressMapping;
+use precompile_utils::{prelude::*, EvmResult};
+use sp_runtime::traits::ConstU32;
+#[cfg(test)]
+mod mock;
+
+#[cfg(test)]
+mod tests;
+
+type GetArrayLimit = ConstU32<ARRAY_LIMIT>;
+
+/// Storage cleaner precompile. This precompile is used to clear the storage of a suicided contract.
+#[derive(Debug, Clone)]
+pub struct StorageCleanerPrecompile<Runtime>(PhantomData<Runtime>);
+
+#[precompile_utils::precompile]
+impl<Runtime> StorageCleanerPrecompile<Runtime>
+where
+	Runtime: pallet_evm::Config,
+{
+	#[precompile::public("clearSuicidedStorage(address[],uint32)")]
+	fn clear_suicided_storage(
+		handle: &mut impl PrecompileHandle,
+		addresses: BoundedVec<Address, GetArrayLimit>,
+		limit: u32,
+	) -> EvmResult {
+		let addresses: Vec<_> = addresses.into();
+		let mut deleted_entries = 0;
+
+		for address in &addresses {
+			// Read Suicided storage item
+			// Suicided: Blake2128(16) + H160(20)
+			handle.record_db_read::<Runtime>(36)?;
+			if !pallet_evm::Pallet::<Runtime>::is_account_suicided(&address.0) {
+				return Err(revert(format!("NotSuicided: {}", address.0)));
+			}
+
```

Comment:
The cost of `clear_suicided_contract`should be accounted here, and we shoudl revert if there is not enough gas for that

Reply:
@librelois  I didn't fully understand. Why should we account for the cost of `clear_suicided_contract` at the beginning?

Reply:
@ahmadkaouk because we must always account for the cost of an operation **before** performing it, and revert if there is not enough remaining gas.

But we can't revert in the loop whiteout calling clear_suicided_contract, so we should account the associated cost before the loop

Reply:
I am making sure to  record the cost for each operation before executing it.

For `clear_suicided_contract`: it removes the address from the list of suicided contracts and reduce the account's sufficients. It should be called only for smart contracts with no remaining entries.  So I cannot call `clear_suicided_contract` before the loop because I need to make sure that all storage entries are deleted before.
(If a contract has numerous storage entries, it may require multiple calls to delete all entries due to limits).

Reply:
@librelois I made the changes so that we record the max cost of the clear_suicided_contract at the beginning before any operation, and refund any extra cost at the end.